src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,250 @@
+    if (!this.permissions) {
+      this.permissions = {};
+    }
+
+    for (const permissionType of permissionTypes) {
+      const result = deleteFromPrincipals(
+        this.permissions![permissionType],
+        principals.users,
+        principals.groups
+      );
+      if (result) {
+        this.permissions[permissionType] = result;
+      }
+    }
+
+    return this;
+  }
+
+  /**
+   * transform permissions format, change the format from permissionType->principals to principal->permissionTypes,
+   * which is used to clearyly dispaly user/group list and their granted permissions in the UI
+   * original permissions:   {
+   *     read: {
+   *         users:['user1']
+   *     },
+   *     write:{
+   *         groups:['group1']
+   *     }
+   * }
+   *
+   * transformed permissions: [
+   *     {type:'users',name:'user1',permissions:['read']},
+   *     {type:'groups',name:'group1',permissions:['write']},
+   * ]
+   */
+  public toFlatList(): TransformedPermission[] {
+    const result: TransformedPermission[] = [];
+    if (!this.permissions) {
+      return result;
+    }
+
+    for (const permissionType in this.permissions) {
+      if (Object.prototype.hasOwnProperty.call(this.permissions, permissionType)) {
+        const { users = [], groups = [] } = this.permissions[permissionType] ?? {};
+        users.forEach((user) => {
+          const found = result.find((r) => r.type === PrincipalType.Users && r.name === user);
+          if (found) {
+            found.permissions.push(permissionType);
+          } else {
+            result.push({ type: PrincipalType.Users, name: user, permissions: [permissionType] });
+          }
+        });
+        groups.forEach((group) => {
+          const found = result.find((r) => r.type === PrincipalType.Groups && r.name === group);
+          if (found) {
+            found.permissions.push(permissionType);
+          } else {
+            result.push({ type: PrincipalType.Groups, name: group, permissions: [permissionType] });
+          }
+        });
+      }
+    }
+
+    return result;
+  }
+
+  public resetPermissions() {
+    // reset permissions
+    this.permissions = {};
+  }
+
+  // return the permissions object
+  public getPermissions() {
+    return this.permissions;
+  }
+
+  /**
+   * generate query DSL by the specific conditions, used for fetching saved objects from the saved objects index
+   */
+  public static generateGetPermittedSavedObjectsQueryDSL(
+    permissionTypes: string[],
+    principals: Principals,
+    savedObjectType?: string | string[]
+  ) {
+    if (!principals || !permissionTypes) {
+      return {
+        query: {
+          match_none: {},
+        },
+      };
+    }
+
+    const bool: any = {
+      filter: [],
+    };
+    const subBool: any = {
+      should: [],
+    };
+
+    permissionTypes.forEach((permissionType) => {
+      Object.entries(principals).forEach(([principalType, principalsInCurrentType]) => {
+        subBool.should.push({
```

Comment:
@joshuarrrr @ashwin-pc do you guys know if this stuff can be built used by the data plugin?

Reply:
not likely in data plugin, since data plugin deals with index data in OpenSearch, while this ACL is for saved objects.